### PR TITLE
Add build strategy to skip events that trigger a build if HEAD commit matches some rules (continued)

### PR DESCRIPTION
**Pull request by mryan43**
This is a follow up on PR #10  by @nfalco79 with changes requested by @witokondoria (kudos to both of them for maintaining/contributing to this plugin !) I have added:
- Github implementation
- Separation of commit author and message filtering into different BranchBuildStrategy classes
- Usage of regex instead of wildcards
- Common logic extracted into parent maven module

I have tested it on both github.com and bitbucket cloud (it should in theory also work with github enterprise and bitbucket server)

Know issue : https://issues.jenkins-ci.org/browse/JENKINS-54468
This only affects users who have both the github and bitbucket flavors of the plugin and there is a simple workaround for affected users (only use the applicable branch build strategy in the UI). I don't think this should be a blocker to merge this PR as the fix would have to be done in the branch-api plugin anyway.

---
This change is [Reviewable](https://reviewable.io/reviews/jenkinsci/scm-trait-commit-skip-plugin/13)

### CODE REVIEW

File: bitbucket-scm-trait-commit-skip/pom.xml [hunk 1]
```diff
@@ -15,18 +15,48 @@
...
+        <repository>
+            <id>incrementals.jenkins-ci.org</id>
+            <url>https://repo.jenkins-ci.org/incrementals/</url>
+        </repository>
+    </repositories>
+
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>com.jcraft</groupId>
```
[the remark this answers is not in the document]
Comment (mryan43) [reply bitbucket-scm-trait-commit-skip/pom.xml]: The bitbucket branch source plugin introduces this enforcer violation all by itself (not in test scope)

> [WARNING] Rule 5: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps failed with message:
Failed while enforcing RequireUpperBoundDeps. The error(s) are [
Require upper bound dependencies error for com.jcraft:jsch:0.1.50 paths to dependency are:
+-org.jenkins-ci.plugins:bitbucket-scm-trait-commit-skip:0.2.0-SNAPSHOT
  +-org.jenkins-ci.plugins:cloudbees-bitbucket-branch-source:2.2.14
    +-org.jenkins-ci.plugins:mercurial:2.0
      +-com.jcraft:jsch:0.1.50
and
+-org.jenkins-ci.plugins:bitbucket-scm-trait-commit-skip:0.2.0-SNAPSHOT
  +-org.jenkins-ci.plugins:cloudbees-bitbucket-branch-source:2.2.14
    +-org.jenkins-ci.plugins:git:3.6.0
      +-org.jenkins-ci.plugins:git-client:2.5.0
        +-org.eclipse.jgit:org.eclipse.jgit:4.5.2.201704071617-r
          +-com.jcraft:jsch:0.1.53

I guess that's why @nfalco79 added this here.

File: github-scm-trait-commit-skip/pom.xml [hunk 2]
```diff
@@ -13,7 +13,17 @@
 
     <name>GitHub Commit Skip SCM Behaviour</name>
     <url>https://wiki.jenkins-ci.org/display/JENKINS/GitHub+Commit+Skip+SCM+Behaviour+Plugin</url>
-    <description>Adds a filtering trait according to the commit message (containing a [ci skip] or [skip ci] text) to an existing Github Branch Source Plugin configuration.</description>
+    <description>Allows to ignore github Branches/Pull Requests/Commits based on the content of commit messages</description>
+
+    <dependencyManagement>
```
[the remark this answers is not in the document]
Comment (mryan43) [reply github-scm-trait-commit-skip/pom.xml]: the github branch source plugin depends transitively on version 1.1, leading ot the following enforcer violation

> [WARNING] Rule 5: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps failed with message:
Failed while enforcing RequireUpperBoundDeps. The error(s) are [
Require upper bound dependencies error for org.jenkins-ci.plugins:plain-credentials:1.1 paths to dependency are:
+-org.jenkins-ci.plugins:github-scm-trait-commit-skip:0.2.0-SNAPSHOT
  +-org.jenkins-ci.plugins:github-branch-source:2.3.6
    +-com.coravy.hudson.plugins.github:github:1.26.0
      +-org.jenkins-ci.plugins:plain-credentials:1.1
and
+-org.jenkins-ci.plugins:github-scm-trait-commit-skip:0.2.0-SNAPSHOT
  +-org.jenkins-ci.plugins.workflow:workflow-aggregator:2.5
    +-org.jenkinsci.plugins:pipeline-model-definition:1.0
      +-org.jenkins-ci.plugins:credentials-binding:1.10
        +-org.jenkins-ci.plugins:plain-credentials:1.3

**Comment by nfalco79** [comment]
@witokondoria the point of dependency management is avoid to exclude or add direct dependencies just to override (manage) the version of transitive dependencies. Workflow-aggregation and workflow-multibranch are sufficient to archive your goal. Other test dependencies are just included to override (and so manage) the version but the correct way is use dependency management because you are not using directly all of those test dependencies.

**Comment by mryan43** [comment]
I found a bug in the github implementation, please wait until I fix it before merging this PR

**Comment by Alex-Vol-SV** [comment]
Excellent work @nfalco79 and @mryan43 

I am interested in this capability and had started my own variation of this PR. I would like to point out a limitation that I found in the implementation. The pattern matching uses space separated lists of patterns which precludes commit message patterns that include spaces. In my variation I used comma separated pattern list to make it work. Escape character for spaces could be an alternative option.

I am also testing against Bitbucket Server and so far this works well. I do not have a Github enterprise instance to report on this functionality for the same.

**Comment by mryan43** [comment]
Hello ! I have removed wildcard patterns support in this PR to simplify the code, only regex is supported, and which solves the whitespace issue you mentionned.

Also note that you can apply the branchbuildstrategy multiple times with different patterns. Does that work arpund the limitation that you mentionned ?

**Comment by Alex-Vol-SV** [comment]
Regex is perfect and one pattern per build strategy application is even better. Great approach, thanks. I am going to pull your current PR to test in my CI, thanks for working on this, my users will appreciate having this feature.

**Comment by nfalco79** [comment]
Like in other trait example in case of regexp should be just one pattern. This have sense because with regexp you can match different patterns.

File: bitbucket-scm-trait-commit-skip/src/main/resources/org/jenkinsci/plugins/scm/trait/commit/skip/Messages.properties [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# The MIT License
```
Comment (nfalco79) [inline bitbucket-scm-trait-commit-skip/src/main/resources/org/jenkinsci/plugins/scm/trait/commit/skip/Messages.properties]: this file is not used anymore

**Comment by mryan43** [comment]
Thanks for the feeback @nfalco79 !

**Comment by michelzanini** [comment]
Very nice work! Waiting for this to get merged and released.

**Comment by michelzanini** [comment]
Can we release a new version now this has being merged? Thanks!

**Comment by onigoetz** [comment]
When will this be released ?

**Comment by witokondoria** [comment]
This is part of the 0.2.0 released plugin. Thanks everyone for your contributions

**Pull request merged**